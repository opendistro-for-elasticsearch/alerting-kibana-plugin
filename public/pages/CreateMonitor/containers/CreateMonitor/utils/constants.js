import { OPERATORS_MAP } from '../../../components/MonitorExpressions/expressions/utils/constants';

/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export const BUCKET_COUNT = 5;

export const MATCH_ALL_QUERY = JSON.stringify({ size: 0, query: { match_all: {} } }, null, 4);

export const FORMIK_INITIAL_VALUES = {
  /* CONFIGURE MONITOR */
  name: '',
  disabled: false,
  frequency: 'interval',
  timezone: [],
  daily: 0,
  period: { interval: 1, unit: 'MINUTES' },
  weekly: { mon: false, tue: false, wed: false, thur: false, fri: false, sat: false, sun: false },
  monthly: { type: 'day', day: 1 },
  cronExpression: '0 */1 * * *',

  /* DEFINE MONITOR */
  searchType: 'graph',
  index: [],
  timeField: '',
  query: MATCH_ALL_QUERY,
  aggregationType: 'count',
  fieldName: [],
  overDocuments: 'all documents',
  groupedOverTop: 5,
  groupedOverFieldName: 'bytes',
  bucketValue: 1,
  bucketUnitOfTime: 'h', // m = minute, h = hour, d = day
  where: {
    fieldName: [],
    operator: OPERATORS_MAP.IS,
    fieldValue: '',
    fieldRangeStart: 0,
    fieldRangeEnd: 0,
  },
  detectorId: '',
};
