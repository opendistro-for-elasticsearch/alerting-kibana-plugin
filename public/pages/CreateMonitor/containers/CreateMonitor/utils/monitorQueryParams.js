import { SEARCH_TYPE } from '../../../../../utils/constants';

export const initializeFromQueryParams = queryParams => {
  return {
    searchType: queryParams.searchType || undefined,
    name: queryParams.name ? `${queryParams.name}-Monitor` : undefined,
    detectorId: queryParams.adId || undefined,
    period:
      queryParams.interval && queryParams.unit
        ? { interval: parseInt(queryParams.interval), unit: queryParams.unit }
        : undefined,
  };
};
