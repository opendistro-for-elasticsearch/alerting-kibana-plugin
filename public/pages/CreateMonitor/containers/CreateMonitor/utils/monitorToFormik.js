/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import { FORMIK_INITIAL_VALUES } from './constants';
import { SEARCH_TYPE, INPUTS_DETECTOR_ID } from '../../../../../utils/constants';

// Convert Monitor JSON to Formik values used in UI forms
export default function monitorToFormik(monitor) {
  const formikValues = _.cloneDeep(FORMIK_INITIAL_VALUES);
  if (!monitor) return formikValues;
  const {
    name,
    enabled,
    schedule: { cron: { expression: cronExpression = formikValues.cronExpression, timezone } = {} },
    inputs,
    ui_metadata: { schedule = {}, search = {} } = {},
  } = monitor;
  // Default searchType to query, because if there is no ui_metadata or search then it was created through API or overwritten by API
  // In that case we don't want to guess on the UI what selections a user made, so we will default to just showing the extraction query
  const { searchType = 'query', fieldName } = search;
  const isAD = searchType === SEARCH_TYPE.AD;

  return {
    /* INITIALIZE WITH DEFAULTS */
    ...formikValues,

    /* CONFIGURE MONITOR */
    name,
    disabled: !enabled,

    /* This will overwrite the fields in use by Monitor from ui_metadata */
    ...schedule,
    cronExpression,

    /* DEFINE MONITOR */
    ...search,
    searchType,
    fieldName: fieldName ? [{ label: fieldName }] : [],
    timezone: timezone ? [{ label: timezone }] : [],

    detectorId: isAD //&& _.get(inputs, '0.search.indices.0') === '.opendistro-anomaly-results*'
      ? _.get(inputs, INPUTS_DETECTOR_ID)
      : undefined,
    index: inputs[0].search.indices.map(index => ({ label: index })),
    query: JSON.stringify(inputs[0].search.query, null, 4),
  };
}
