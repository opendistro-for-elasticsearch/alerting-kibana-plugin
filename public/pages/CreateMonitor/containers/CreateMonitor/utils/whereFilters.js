/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { OPERATORS_MAP } from '../../../components/MonitorExpressions/expressions/utils/constants';
import { DATA_TYPES } from '../../../../../utils/constants';

//TODO:: Breakdown to factory pattern for rules in-case we support multiple filters. This is just ease for the single one
export const OPERATORS_QUERY_MAP = {
  [OPERATORS_MAP.IS]: {
    query: ({ fieldName: [{ label, type }], fieldValue }) =>
      type === DATA_TYPES.TEXT
        ? { match_phrase: { [label]: fieldValue } }
        : { term: { [label]: fieldValue } },
  },
  [OPERATORS_MAP.IS_NOT]: {
    query: ({ fieldName: [{ label, type }], fieldValue }) =>
      type === DATA_TYPES.TEXT
        ? {
            bool: { must_not: { match_phrase: { [label]: fieldValue } } },
          }
        : {
            bool: { must_not: { term: { [label]: fieldValue } } },
          },
  },
  [OPERATORS_MAP.IS_NULL]: {
    query: ({ fieldName: [{ label: fieldKey }] }) => ({
      bool: { must_not: { exists: { field: fieldKey } } },
    }),
  },
  [OPERATORS_MAP.IS_NOT_NULL]: {
    query: ({ fieldName: [{ label: fieldKey }] }) => ({ exists: { field: fieldKey } }),
  },
  [OPERATORS_MAP.IS_GREATER]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      range: { [fieldKey]: { gt: fieldValue } },
    }),
  },

  [OPERATORS_MAP.IS_GREATER_EQUAL]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      range: { [fieldKey]: { gte: fieldValue } },
    }),
  },
  [OPERATORS_MAP.IS_LESS]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      range: { [fieldKey]: { lt: fieldValue } },
    }),
  },

  [OPERATORS_MAP.IS_LESS_EQUAL]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      range: { [fieldKey]: { lte: fieldValue } },
    }),
  },

  [OPERATORS_MAP.IN_RANGE]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldRangeStart, fieldRangeEnd }) => ({
      range: { [fieldKey]: { gte: fieldRangeStart, lte: fieldRangeEnd } },
    }),
  },
  [OPERATORS_MAP.NOT_IN_RANGE]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldRangeStart, fieldRangeEnd }) => ({
      bool: { must_not: { range: { [fieldKey]: { gte: fieldRangeStart, lte: fieldRangeEnd } } } },
    }),
  },

  [OPERATORS_MAP.STARTS_WITH]: {
    query: ({ fieldName, fieldValue }) => ({ prefix: { [fieldName[0].label]: fieldValue } }),
  },

  [OPERATORS_MAP.ENDS_WITH]: {
    query: ({ fieldName, fieldValue }) => ({
      wildcard: { [fieldName[0].label]: `*${fieldValue}` },
    }),
  },
  [OPERATORS_MAP.CONTAINS]: {
    query: ({ fieldName: [{ label, type }], fieldValue }) =>
      type === DATA_TYPES.TEXT
        ? {
            query_string: { query: `*${fieldValue}*`, default_field: label },
          }
        : {
            wildcard: { [label]: `*${fieldValue}*` },
          },
  },
  [OPERATORS_MAP.NOT_CONTAINS]: {
    query: ({ fieldName: [{ label, type }], fieldValue }) =>
      type === DATA_TYPES.TEXT
        ? {
            bool: {
              must_not: { query_string: { query: `*${fieldValue}*`, default_field: label } },
            },
          }
        : {
            bool: { must_not: { wildcard: { [label]: `*${fieldValue}*` } } },
          },
  },
};
