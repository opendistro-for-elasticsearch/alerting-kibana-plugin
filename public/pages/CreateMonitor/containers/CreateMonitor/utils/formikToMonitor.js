/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import moment from 'moment-timezone';
import { BUCKET_COUNT } from './constants';
import { SEARCH_TYPE } from '../../../../../utils/constants';
import { OPERATORS_QUERY_MAP } from './whereFilters';

export function formikToMonitor(values) {
  const uiSchedule = formikToUiSchedule(values);
  const schedule = buildSchedule(values.frequency, uiSchedule);
  const uiSearch = formikToUiSearch(values);
  return {
    name: values.name,
    type: 'monitor',
    enabled: !values.disabled,
    schedule,
    inputs: [formikToSearch(values)],
    triggers: [],
    ui_metadata: {
      schedule: uiSchedule,
      search: uiSearch,
    },
  };
}

export function formikToInputs(values) {
  const isAD = values.searchType === SEARCH_TYPE.AD;
  return [isAD ? formikToAd(values) : formikToSearch(values)];
}

export function formikToSearch(values) {
  const isAD = values.searchType === SEARCH_TYPE.AD;
  let query = isAD ? formikToAdQuery(values) : formikToQuery(values);
  const indices = isAD ? ['.opendistro-anomaly-results*'] : formikToIndices(values);

  return {
    search: {
      indices,
      query,
    },
  };
}

export function formikToAdQuery(values) {
  return {
    size: 1,
    sort: [{ anomaly_grade: 'desc' }, { confidence: 'desc' }],
    query: {
      bool: {
        filter: [
          {
            range: {
              data_end_time: {
                from: '{{period_end}}||-' + values.period.interval + 'm',
                to: '{{period_end}}',
                include_lower: true,
                include_upper: true,
              },
            },
          },
          {
            term: {
              detector_id: {
                value: values.detectorId,
              },
            },
          },
        ],
      },
    },
    aggregations: {
      max_anomaly_grade: {
        max: {
          field: 'anomaly_grade',
        },
      },
    },
  };
}

export function formikToAd(values) {
  return {
    anomaly_detector: {
      detector_id: values.detectorId,
    },
  };
}
export function formikToUiSearch(values) {
  const {
    searchType,
    aggregationType,
    timeField,
    fieldName: [{ label: fieldName = '' } = {}],
    overDocuments,
    groupedOverTop,
    groupedOverFieldName,
    bucketValue,
    bucketUnitOfTime,
    where,
  } = values;
  return {
    searchType,
    aggregationType,
    timeField,
    fieldName,
    overDocuments,
    groupedOverTop,
    groupedOverFieldName,
    bucketValue,
    bucketUnitOfTime,
    where,
  };
}

export function formikToIndices(values) {
  return values.index.map(({ label }) => label);
}

export function formikToQuery(values) {
  const isGraph = values.searchType === SEARCH_TYPE.GRAPH;
  return isGraph ? formikToGraphQuery(values) : formikToExtractionQuery(values);
}

export function formikToExtractionQuery(values) {
  return JSON.parse(values.query);
}

export function formikToGraphQuery(values) {
  const { bucketValue, bucketUnitOfTime } = values;
  const whenAggregation = formikToWhenAggregation(values);
  const timeField = values.timeField;
  const filters = [
    {
      range: {
        [timeField]: {
          gte: `{{period_end}}||-${Math.round(bucketValue)}${bucketUnitOfTime}`,
          lte: '{{period_end}}',
          format: 'epoch_millis',
        },
      },
    },
  ];
  const whereClause = formikToWhereClause(values);
  if (whereClause) {
    filters.push({ ...whereClause });
  }
  return {
    size: 0,
    aggregations: whenAggregation,
    query: {
      bool: {
        filter: filters,
      },
    },
  };
}

export function formikToUiGraphQuery(values) {
  const { bucketValue, bucketUnitOfTime } = values;
  const overAggregation = formikToUiOverAggregation(values);
  const timeField = values.timeField;
  const filters = [
    {
      range: {
        [timeField]: {
          // default range window to [BUCKET_COUNT] * the date histogram interval
          gte: `now-${bucketValue * BUCKET_COUNT}${bucketUnitOfTime}`,
          lte: 'now',
        },
      },
    },
  ];
  const whereClause = formikToWhereClause(values);
  if (whereClause) {
    filters.push({ ...whereClause });
  }
  return {
    size: 0,
    aggregations: overAggregation,
    query: {
      bool: {
        filter: filters,
      },
    },
  };
}

export function formikToUiOverAggregation(values) {
  const whenAggregation = formikToWhenAggregation(values);
  const { bucketValue, bucketUnitOfTime } = values;
  const timeField = values.timeField;

  return {
    over: {
      date_histogram: {
        field: timeField,
        interval: `${bucketValue}${bucketUnitOfTime}`,
        time_zone: moment.tz.guess(),
        min_doc_count: 0,
        extended_bounds: {
          min: `now-${bucketValue * BUCKET_COUNT}${bucketUnitOfTime}`,
          max: 'now',
        },
      },
      aggregations: whenAggregation,
    },
  };
}

export function formikToWhereClause({ where }) {
  if (where.fieldName.length > 0) {
    return OPERATORS_QUERY_MAP[where.operator].query(where);
  }
}

export function formikToWhenAggregation(values) {
  const {
    aggregationType,
    fieldName: [{ label: field } = {}],
  } = values;
  if (aggregationType === 'count' || !field) return {};
  return { when: { [aggregationType]: { field } } };
}

export function formikToUiSchedule(values) {
  return {
    timezone: _.get(values, 'timezone[0].label', null),
    frequency: values.frequency,
    period: values.period,
    daily: values.daily,
    weekly: values.weekly,
    monthly: values.monthly,
    cronExpression: values.cronExpression,
  };
}

export function buildSchedule(scheduleType, values) {
  const {
    period,
    daily,
    weekly,
    monthly: { type, day },
    cronExpression,
    timezone,
  } = values;
  switch (scheduleType) {
    case 'interval': {
      return { period };
    }
    case 'daily': {
      return { cron: { expression: `0 ${daily} * * *`, timezone } };
    }
    case 'weekly': {
      const daysOfWeek = Object.entries(weekly)
        .filter(([day, checked]) => checked)
        .map(([day]) => day.toUpperCase())
        .join(',');
      return { cron: { expression: `0 ${daily} * * ${daysOfWeek}`, timezone } };
    }
    case 'monthly': {
      let dayOfMonth = '?';
      if (type === 'day') {
        dayOfMonth = day;
      }
      return { cron: { expression: `0 ${daily} ${dayOfMonth} */1 *`, timezone } };
    }
    case 'cronExpression':
      return { cron: { expression: cronExpression, timezone } };
  }
}
