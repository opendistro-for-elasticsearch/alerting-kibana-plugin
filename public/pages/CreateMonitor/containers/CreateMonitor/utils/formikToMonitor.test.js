/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import {
  formikToMonitor,
  formikToUiSearch,
  formikToIndices,
  formikToQuery,
  formikToExtractionQuery,
  formikToGraphQuery,
  formikToUiGraphQuery,
  formikToUiOverAggregation,
  formikToWhenAggregation,
  formikToUiSchedule,
  buildSchedule,
  formikToWhereClause,
  formikToAd,
} from './formikToMonitor';

import { FORMIK_INITIAL_VALUES } from './constants';
import { OPERATORS_MAP } from '../../../components/MonitorExpressions/expressions/utils/constants';

jest.mock('moment-timezone', () => {
  const moment = require.requireActual('moment-timezone');
  moment.tz.guess = () => 'America/Los_Angeles';
  return moment;
});

describe('formikToMonitor', () => {
  const formikValues = _.cloneDeep(FORMIK_INITIAL_VALUES);
  formikValues.name = 'random_name';
  formikValues.disabled = true;
  formikValues.index = [{ label: 'index1' }, { label: 'index2' }];
  formikValues.fieldName = [{ label: 'bytes' }];
  formikValues.timezone = [{ label: 'America/Los_Angeles' }];
  test('can build monitor', () => {
    expect(formikToMonitor(formikValues)).toMatchSnapshot();
  });
});

describe('formikToDetector', () => {
  const formikValues = _.cloneDeep(FORMIK_INITIAL_VALUES);
  formikValues.detectorId = 'temp_detector';
  test('can build detector', () => {
    expect(formikToAd(formikValues)).toMatchSnapshot();
  });
});

describe('formikToUiSearch', () => {
  const formikValues = _.cloneDeep(FORMIK_INITIAL_VALUES);
  formikValues.fieldName = [{ label: 'bytes' }];
  formikValues.timeField = '@timestamp';
  test('can build ui search', () => {
    expect(formikToUiSearch(formikValues)).toMatchSnapshot();
  });
  test('can build ui search with term where field', () => {
    formikValues.where = {
      fieldName: [{ label: 'age', type: 'number' }],
      operator: OPERATORS_MAP.IS_GREATER_EQUAL,
      fieldValue: 20,
    };
    expect(formikToUiSearch(formikValues)).toMatchSnapshot();
  });

  test('can build ui search with range where field', () => {
    formikValues.where = {
      fieldName: [{ label: 'age', type: 'number' }],
      operator: OPERATORS_MAP.IN_RANGE,
      fieldRangeStart: 20,
      fieldRangeEnd: 40,
    };
    expect(formikToUiSearch(formikValues)).toMatchSnapshot();
  });
});

describe('formikToIndices', () => {
  test('can build index', () => {
    const formikValues = _.cloneDeep(FORMIK_INITIAL_VALUES);
    formikValues.index = [{ label: 'index1' }, { label: 'index2' }];
    expect(formikToIndices(formikValues)).toMatchSnapshot();
  });
});

describe('formikToQuery', () => {
  const formikValues = _.cloneDeep(FORMIK_INITIAL_VALUES);

  test('can build graph query', () => {
    expect(formikToQuery({ ...formikValues, timeField: '@timestamp' })).toMatchSnapshot();
  });

  test('can build extraction query', () => {
    formikValues.searchType = 'query';
    expect(formikToQuery(formikValues)).toMatchSnapshot();
  });
});

describe('formikToExtractionQuery', () => {
  test('can extract query', () => {
    expect(formikToExtractionQuery(FORMIK_INITIAL_VALUES)).toMatchSnapshot();
  });
});

describe('formikToGraphQuery', () => {
  test('can build graph query', () => {
    expect(
      formikToGraphQuery({ ...FORMIK_INITIAL_VALUES, timeField: '@timestamp' })
    ).toMatchSnapshot();
  });
});

describe('formikToUiGraphQuery', () => {
  test('can build ui graph query', () => {
    expect(
      formikToUiGraphQuery({ ...FORMIK_INITIAL_VALUES, timeField: '@timestamp' })
    ).toMatchSnapshot();
  });
});

describe('formikToUiOverAggregation', () => {
  test('can build over aggregation', () => {
    expect(
      formikToUiOverAggregation({ ...FORMIK_INITIAL_VALUES, timeField: '@timestamp' })
    ).toMatchSnapshot();
  });
});

describe('formikToWhenAggregation', () => {
  const formikValues = _.cloneDeep(FORMIK_INITIAL_VALUES);

  test('can build when (count) aggregation', () => {
    expect(formikToWhenAggregation(formikValues)).toMatchSnapshot();
  });

  test('can build when aggregation', () => {
    formikValues.aggregationType = 'avg';
    formikValues.fieldName = [{ label: 'bytes' }];
    expect(formikToWhenAggregation(formikValues)).toMatchSnapshot();
  });
});

describe('formikToUiSchedule', () => {
  test('can build uiSchedule', () => {
    expect(
      formikToUiSchedule({ ...FORMIK_INITIAL_VALUES, timezone: [{ label: 'America/Los_Angeles' }] })
    ).toMatchSnapshot();
  });
});

describe('buildSchedule', () => {
  let formikValues;
  let uiSchedule;
  beforeEach(() => {
    formikValues = _.cloneDeep(FORMIK_INITIAL_VALUES);
    formikValues.timezone = [{ label: 'America/Los_Angeles' }];
    uiSchedule = formikToUiSchedule(formikValues);
  });

  test('can build interval schedule', () => {
    expect(buildSchedule('interval', uiSchedule)).toMatchSnapshot();
  });

  test('can build daily schedule', () => {
    expect(buildSchedule('daily', uiSchedule)).toMatchSnapshot();
  });

  test('can build weekly schedule', () => {
    uiSchedule.weekly.tue = true;
    uiSchedule.weekly.thur = true;
    expect(buildSchedule('weekly', uiSchedule)).toMatchSnapshot();
  });

  test('can build monthly (day) schedule', () => {
    uiSchedule.monthly.type = 'day';
    expect(buildSchedule('monthly', uiSchedule)).toMatchSnapshot();
  });

  test('can build cron schedule', () => {
    expect(buildSchedule('cronExpression', uiSchedule)).toMatchSnapshot();
  });
});

describe('formikToWhereClause', () => {
  const numericFieldName = [{ label: 'age', type: 'number' }];
  const textField = [{ label: 'city', type: 'text' }];
  const keywordField = [{ label: 'city.keyword', type: 'keyword' }];

  test.each([
    [numericFieldName, OPERATORS_MAP.IS, 20, { term: { age: 20 } }],
    [textField, OPERATORS_MAP.IS, 'Seattle', { match_phrase: { city: 'Seattle' } }],
    [numericFieldName, OPERATORS_MAP.IS_NOT, 20, { bool: { must_not: { term: { age: 20 } } } }],
    [
      textField,
      OPERATORS_MAP.IS_NOT,
      'Seattle',
      { bool: { must_not: { match_phrase: { city: 'Seattle' } } } },
    ],
    [
      numericFieldName,
      OPERATORS_MAP.IS_NULL,
      undefined,
      { bool: { must_not: { exists: { field: 'age' } } } },
    ],
    [numericFieldName, OPERATORS_MAP.IS_NOT_NULL, undefined, { exists: { field: 'age' } }],
    [numericFieldName, OPERATORS_MAP.IS_GREATER, 20, { range: { age: { gt: 20 } } }],
    [numericFieldName, OPERATORS_MAP.IS_GREATER_EQUAL, 20, { range: { age: { gte: 20 } } }],
    [numericFieldName, OPERATORS_MAP.IS_LESS, 20, { range: { age: { lt: 20 } } }],
    [numericFieldName, OPERATORS_MAP.IS_LESS_EQUAL, 20, { range: { age: { lte: 20 } } }],
    [textField, OPERATORS_MAP.STARTS_WITH, 'Se', { prefix: { city: 'Se' } }],
    [textField, OPERATORS_MAP.ENDS_WITH, 'Se', { wildcard: { city: '*Se' } }],
    [
      textField,
      OPERATORS_MAP.CONTAINS,
      'Se',
      { query_string: { query: `*Se*`, default_field: 'city' } },
    ],
    [keywordField, OPERATORS_MAP.CONTAINS, 'Se', { wildcard: { 'city.keyword': '*Se*' } }],
    [
      textField,
      OPERATORS_MAP.NOT_CONTAINS,
      'Se',
      { bool: { must_not: { query_string: { query: `*Se*`, default_field: 'city' } } } },
    ],
  ])('.formikToWhereClause (%j,  %S)', (fieldName, operator, fieldValue, expected) => {
    expect(formikToWhereClause({ where: { fieldName, operator, fieldValue } })).toEqual(expected);
  });
});
