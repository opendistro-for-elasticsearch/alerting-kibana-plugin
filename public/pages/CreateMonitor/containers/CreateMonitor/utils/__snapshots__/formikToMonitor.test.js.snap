// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildSchedule can build cron schedule 1`] = `
Object {
  "cron": Object {
    "expression": "0 */1 * * *",
    "timezone": "America/Los_Angeles",
  },
}
`;

exports[`buildSchedule can build daily schedule 1`] = `
Object {
  "cron": Object {
    "expression": "0 0 * * *",
    "timezone": "America/Los_Angeles",
  },
}
`;

exports[`buildSchedule can build interval schedule 1`] = `
Object {
  "period": Object {
    "interval": 1,
    "unit": "MINUTES",
  },
}
`;

exports[`buildSchedule can build monthly (day) schedule 1`] = `
Object {
  "cron": Object {
    "expression": "0 0 1 */1 *",
    "timezone": "America/Los_Angeles",
  },
}
`;

exports[`buildSchedule can build weekly schedule 1`] = `
Object {
  "cron": Object {
    "expression": "0 0 * * TUE,THUR",
    "timezone": "America/Los_Angeles",
  },
}
`;

exports[`formikToDetector can build detector 1`] = `
Object {
  "anomaly_detector": Object {
    "detector_id": "temp_detector",
  },
}
`;

exports[`formikToExtractionQuery can extract query 1`] = `
Object {
  "query": Object {
    "match_all": Object {},
  },
  "size": 0,
}
`;

exports[`formikToGraphQuery can build graph query 1`] = `
Object {
  "aggregations": Object {},
  "query": Object {
    "bool": Object {
      "filter": Array [
        Object {
          "range": Object {
            "@timestamp": Object {
              "format": "epoch_millis",
              "gte": "{{period_end}}||-1h",
              "lte": "{{period_end}}",
            },
          },
        },
      ],
    },
  },
  "size": 0,
}
`;

exports[`formikToIndices can build index 1`] = `
Array [
  "index1",
  "index2",
]
`;

exports[`formikToMonitor can build monitor 1`] = `
Object {
  "enabled": false,
  "inputs": Array [
    Object {
      "search": Object {
        "indices": Array [
          "index1",
          "index2",
        ],
        "query": Object {
          "aggregations": Object {},
          "query": Object {
            "bool": Object {
              "filter": Array [
                Object {
                  "range": Object {
                    "": Object {
                      "format": "epoch_millis",
                      "gte": "{{period_end}}||-1h",
                      "lte": "{{period_end}}",
                    },
                  },
                },
              ],
            },
          },
          "size": 0,
        },
      },
    },
  ],
  "name": "random_name",
  "schedule": Object {
    "period": Object {
      "interval": 1,
      "unit": "MINUTES",
    },
  },
  "triggers": Array [],
  "type": "monitor",
  "ui_metadata": Object {
    "schedule": Object {
      "cronExpression": "0 */1 * * *",
      "daily": 0,
      "frequency": "interval",
      "monthly": Object {
        "day": 1,
        "type": "day",
      },
      "period": Object {
        "interval": 1,
        "unit": "MINUTES",
      },
      "timezone": "America/Los_Angeles",
      "weekly": Object {
        "fri": false,
        "mon": false,
        "sat": false,
        "sun": false,
        "thur": false,
        "tue": false,
        "wed": false,
      },
    },
    "search": Object {
      "aggregationType": "count",
      "bucketUnitOfTime": "h",
      "bucketValue": 1,
      "fieldName": "bytes",
      "groupedOverFieldName": "bytes",
      "groupedOverTop": 5,
      "overDocuments": "all documents",
      "searchType": "graph",
      "timeField": "",
      "where": Object {
        "fieldName": Array [],
        "fieldRangeEnd": 0,
        "fieldRangeStart": 0,
        "fieldValue": "",
        "operator": "is",
      },
    },
  },
}
`;

exports[`formikToQuery can build extraction query 1`] = `
Object {
  "query": Object {
    "match_all": Object {},
  },
  "size": 0,
}
`;

exports[`formikToQuery can build graph query 1`] = `
Object {
  "aggregations": Object {},
  "query": Object {
    "bool": Object {
      "filter": Array [
        Object {
          "range": Object {
            "@timestamp": Object {
              "format": "epoch_millis",
              "gte": "{{period_end}}||-1h",
              "lte": "{{period_end}}",
            },
          },
        },
      ],
    },
  },
  "size": 0,
}
`;

exports[`formikToUiGraphQuery can build ui graph query 1`] = `
Object {
  "aggregations": Object {
    "over": Object {
      "aggregations": Object {},
      "date_histogram": Object {
        "extended_bounds": Object {
          "max": "now",
          "min": "now-5h",
        },
        "field": "@timestamp",
        "interval": "1h",
        "min_doc_count": 0,
        "time_zone": "America/Los_Angeles",
      },
    },
  },
  "query": Object {
    "bool": Object {
      "filter": Array [
        Object {
          "range": Object {
            "@timestamp": Object {
              "gte": "now-5h",
              "lte": "now",
            },
          },
        },
      ],
    },
  },
  "size": 0,
}
`;

exports[`formikToUiOverAggregation can build over aggregation 1`] = `
Object {
  "over": Object {
    "aggregations": Object {},
    "date_histogram": Object {
      "extended_bounds": Object {
        "max": "now",
        "min": "now-5h",
      },
      "field": "@timestamp",
      "interval": "1h",
      "min_doc_count": 0,
      "time_zone": "America/Los_Angeles",
    },
  },
}
`;

exports[`formikToUiSchedule can build uiSchedule 1`] = `
Object {
  "cronExpression": "0 */1 * * *",
  "daily": 0,
  "frequency": "interval",
  "monthly": Object {
    "day": 1,
    "type": "day",
  },
  "period": Object {
    "interval": 1,
    "unit": "MINUTES",
  },
  "timezone": "America/Los_Angeles",
  "weekly": Object {
    "fri": false,
    "mon": false,
    "sat": false,
    "sun": false,
    "thur": false,
    "tue": false,
    "wed": false,
  },
}
`;

exports[`formikToUiSearch can build ui search 1`] = `
Object {
  "aggregationType": "count",
  "bucketUnitOfTime": "h",
  "bucketValue": 1,
  "fieldName": "bytes",
  "groupedOverFieldName": "bytes",
  "groupedOverTop": 5,
  "overDocuments": "all documents",
  "searchType": "graph",
  "timeField": "@timestamp",
  "where": Object {
    "fieldName": Array [],
    "fieldRangeEnd": 0,
    "fieldRangeStart": 0,
    "fieldValue": "",
    "operator": "is",
  },
}
`;

exports[`formikToUiSearch can build ui search with range where field 1`] = `
Object {
  "aggregationType": "count",
  "bucketUnitOfTime": "h",
  "bucketValue": 1,
  "fieldName": "bytes",
  "groupedOverFieldName": "bytes",
  "groupedOverTop": 5,
  "overDocuments": "all documents",
  "searchType": "graph",
  "timeField": "@timestamp",
  "where": Object {
    "fieldName": Array [
      Object {
        "label": "age",
        "type": "number",
      },
    ],
    "fieldRangeEnd": 40,
    "fieldRangeStart": 20,
    "operator": "in_range",
  },
}
`;

exports[`formikToUiSearch can build ui search with term where field 1`] = `
Object {
  "aggregationType": "count",
  "bucketUnitOfTime": "h",
  "bucketValue": 1,
  "fieldName": "bytes",
  "groupedOverFieldName": "bytes",
  "groupedOverTop": 5,
  "overDocuments": "all documents",
  "searchType": "graph",
  "timeField": "@timestamp",
  "where": Object {
    "fieldName": Array [
      Object {
        "label": "age",
        "type": "number",
      },
    ],
    "fieldValue": 20,
    "operator": "is_greater_equal",
  },
}
`;

exports[`formikToWhenAggregation can build when (count) aggregation 1`] = `Object {}`;

exports[`formikToWhenAggregation can build when aggregation 1`] = `
Object {
  "when": Object {
    "avg": Object {
      "field": "bytes",
    },
  },
}
`;
